### PR TITLE
<iomanip>: get_time handles leading zeros incorrectly

### LINKED ISSUE #1606: <iomanip>: get_time handles leading zeros incorrectly
Reported by @randydu in https://github.com/microsoft/STL/issues/944#issuecomment-771272860 (thanks @statementreply for looking into this).

While reviewing #1280, we noticed and fixed an issue with leading zeros, but the fix was incomplete. Here's a slightly enhanced version of @randydu's test case:

```
C:\Temp>type meow.cpp
```
```cpp
#include <cassert>
#include <cstdio>
#include <ctime>
#include <iomanip>
#include <sstream>
using namespace std;

int main() {
  istringstream iss("19700405T000006");
  tm t{};
  iss >> get_time(&t, "%Y%m%dT%H%M%S");
  assert(iss);

  printf("Expected hour 0, min 0, sec 6\n");
  printf("     Got hour %d, min %d, sec %d\n", t.tm_hour, t.tm_min, t.tm_sec);

  assert(t.tm_year == 70);
  assert(t.tm_mon == 3);
  assert(t.tm_mday == 5);
  assert(t.tm_hour == 0);
  assert(t.tm_min == 0);
  assert(t.tm_sec == 6);
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /MTd meow.cpp
meow.cpp

C:\Temp>meow
Expected hour 0, min 0, sec 6
     Got hour 0, min 6, sec 0
Assertion failed: t.tm_min == 0, file meow.cpp, line 21
```

We start by calculating the maximum number of digits we can consume (currently always within [1, 4]):
https://github.com/microsoft/STL/blob/6779b613e11ef9982d4d1e5cca24296f8aecc374/stl/inc/xloctime#L543

But we'll consume an unlimited number of leading zeros, updating `_Digits_seen` accordingly:
https://github.com/microsoft/STL/blob/6779b613e11ef9982d4d1e5cca24296f8aecc374/stl/inc/xloctime#L561-L563

When we reach the point of consuming digits after the leading zeros, we check `_Digits_seen < _Hi_digits`:
https://github.com/microsoft/STL/blob/6779b613e11ef9982d4d1e5cca24296f8aecc374/stl/inc/xloctime#L569-L571

So in this test case, the hour field `%H` consumes `00000` (after the `T`), then notices it's consumed way more than the limit of 2 digits for an hour. It succeeds, and `t.tm_hour == 0`, but we should have consumed only `00` to get that result.

Then, the minute field `%M` consumes the `6` at the end of the input.

We end up returning with only `eofbit` set, and never parse the `%S`. (Separately, I am surprised that we don't `failbit` when we were unable to parse all fields, but I am not an iostreams expert - this part may be intentional.)

https://github.com/microsoft/STL/blob/6779b613e11ef9982d4d1e5cca24296f8aecc374/stl/inc/xloctime#L134-L139

I believe that the leading zeros loop needs to be limited with `_Digits_seen < _Hi_digits`. I'm unsure whether other code changes will be necessary.

### PR DESCRIPTION
**Pull request by MattStephanson**
Fixes #1606. Also:
 - failbit should be set if the stream is EOF while not parsing a specifier.
 - A '%' at the end of a format string is an error, not a literal.
 - Reads leading spaces, in addition to leading zeros.
 - Consistently allow 60 for seconds field.

### CODE REVIEW

**Comment by MattStephanson** [comment]
In https://github.com/microsoft/STL/issues/1606#issue-798894105, @StephanTLavavej noted
> Separately, I am surprised that we don't failbit when we were unable to parse all fields, but I am not an iostreams expert - this part may be intentional.

This appears to be the product of a perhaps-unintended interaction between [[locale.time.get.members]/8](http://eel.is/c++draft/locale.time.get#members-8) and [[locale.time.get.virtuals]/12](http://eel.is/c++draft/locale.time.get#virtuals-12). The later requires only `eofbit` to be set if the stream ends while parsing a specifier in `time_get::do_get`. Afterwards, back in `time_get::get`, the test for `err == ios_­base​::​goodbit` is before the test for `s == end`, and the loop exits without setting any additional flags. So if the EOF happens in `do_get`, only `eofbit` is set. In contrast, if EOF occurs in `get` while parsing a literal or whitespace, then `failbit` is also set. I'm not a fan of this inconsistency. Maybe an LWG Issue would be worthwhile?

File: stl/inc/xloctime [hunk 1]
```diff
@@ -93,37 +93,38 @@ public:
         _State = ios_base::goodbit;
 
         for (; _Fmtfirst != _Fmtlast; ++_Fmtfirst) {
-            if (_Ctype_fac.narrow(*_Fmtfirst) != '%') { // match literal element
+            if (_State != ios_base::goodbit) {
+                // [locale.time.get.members] p8.2
```
Comment (StephanTLavavej) [inline stl/inc/xloctime:97]: We try to cite the Working Paper in addition to the stable name and paragraph, because paragraphs are commonly renumbered (and wording sometimes moves between stable names).

(Applies to the other citations.)

File: stl/inc/xloctime [hunk 2]
```diff
@@ -93,37 +93,38 @@ public:
...
+                // p8.3
+                _State = ios_base::eofbit | ios_base::failbit;
+                break;
+            } else if (_Ctype_fac.narrow(*_Fmtfirst) != '%') { // match literal element
                 if (_Ctype_fac.is(_Ctype::space, *_Fmtfirst)) {
                     while (_First != _Last && _Ctype_fac.is(_Ctype::space, *_First)) {
                         ++_First;
                     }
-                } else if (*_First != *_Fmtfirst) { // bad literal match
+                } else if (_Ctype_fac.toupper(*_First) != _Ctype_fac.toupper(*_Fmtfirst)) { // bad literal match
```
Comment (StephanTLavavej) [inline stl/inc/xloctime:109]: I see that WG21-N4878 [locale.time.get.members]/9 says:

> [Note 1: The function uses the `ctype<charT>` facet installed in `f`'s locale to determine valid whitespace characters. It is unspecified by what means the function performs case-insensitive comparison or whether multi-character sequences are considered while doing so. — end note]

While it's unspecified, we've used `tolower` elsewhere; I think we should be consistent here:

https://github.com/microsoft/STL/blob/30ac4400e528e9837f2e8d9867da668b0de3f7eb/stl/inc/xlocale#L3280-L3282

**Review by StephanTLavavej — COMMENTED** [review]
Thanks! I think I'm following all of the logic here (I especially like how it mirrors the Standardese more closely now), and the added test coverage is great (along with the fact that none of the existing tests are affected). I was the most curious about counting spaces as digits, but that appears to handle the final test case and the `strptime` wording. I'll push a couple of commits for WP citations and consistent case-insensitive comparisons, then I think this is ready for final review.

**Review by StephanTLavavej — APPROVED** [review]

**Review by mnatsuhara — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for fixing these bugs - parsing times without delimiters has been a top customer request! :tada: :date: :watch: :smiley_cat:
